src/main/scala/org/constellation/infrastructure/endpoints/SnapshotEndpoints.scala

```diff
@@ -116,18 +119,33 @@ class SnapshotEndpoints[F[_]](implicit F: Concurrent[F], C: ContextShift[F]) ext
         redownloadService.getAcceptedSnapshots().map(_.asJson).flatMap(Ok(_))
     }
 
-  private def getPeerProposals(nodeId: Id, redownloadService: RedownloadService[F]): HttpRoutes[F] =
+  private def getPeerProposals(selfId: Id, redownloadService: RedownloadService[F]): HttpRoutes[F] =
     HttpRoutes.of[F] {
       case GET -> Root / "peer" / peerId / "snapshot" / "created" =>
         val peerProposals =
-          if (Id(peerId) == nodeId)
+          if (Id(peerId) == selfId)
             redownloadService.getCreatedSnapshots()
           else
-            redownloadService.getPeerProposals(nodeId).map(_.getOrElse(Map.empty))
+            redownloadService.getPeerProposals(Id(peerId)).map(_.getOrElse(Map.empty))
 
         peerProposals.map(_.asJson).flatMap(Ok(_))
     }
 
+  private def queryPeerProposals(selfId: Id, redownloadService: RedownloadService[F]): HttpRoutes[F] =
+    HttpRoutes.of[F] {
+      case req @ POST -> Root / "snapshot" / "proposal" / "query" =>
+        for {
+          query <- req.decodeJson[List[(Id, Long)]]
+          result <- query.traverse {
+            case (`selfId`, height) =>
+              redownloadService.getCreatedSnapshots().map(_.get(height))
+            case (peerId, height) =>
+              redownloadService.getPeerProposals(peerId).map(_.flatMap(_.get(height)))
+          }
```

Comment:
Is there a possibility that getCreatedSnapshots or getPeerProposals will throw hence breaking the whole traversal?

Reply:
It's a read on a map, I think it should never throw.